### DISCUSSION TITLE
Improve BG Subtraction

### DISCUSSION THREAD

**Discussion by gpete** [blakeblackshear/frigate#1288] [Ideas]
I'm finding that objects that move with the wind triggers object detection on an object adjacent to it in the frame which is actually stationary. (i.e. parked car behind a blowing tree)

In looking into opencv motion detection I came across a feature that looks pretty nice and has the possibility to fix a few issues with false positives if used correctly: [Background Subtraction](https://docs.opencv.org/master/d1/dc5/tutorial_background_subtraction.html)

A background model is initialized and then used to subtract objects that are already part of the scene. The current frame is then compared to it and what has changed beyond a certain threshold is considered part of the "foreground". The background then needs to be updated to adapt to scene changes (car parks in front of your camera, initially counts as motion, but eventually becomes a stationary object in the background).

I'm hoping this could even be applied to a tree and shrubs that are shifting slightly. After seeing this I'm now pretty certain it is exactly how the composite "summary" is done where motion events are superimposed (https://v.redd.it/flfjtitwfyd31, mentioned in issue #54). It may also offer a solution to ignoring stationary objects (#841) since objects that aren't moving will eventually become part of the background mask.

I've been trying to investigate ways to reduce false positives on my setup including my ongoing work to train my own custom model for better detection. I think I'm going to shift my focus to this because it should offer a better starting point for object detection.

**Comment by gpete** [comment]
I've spent a few hours looking into this more. Frigate already does a form of background subtraction. I'm not sure what algorithm it is (it doesn't use the OpenCV library for it) but it is fast and effective. My research was focused on my main use-case/want is to ignore the noise of trees and bushes that move and blow around for hours wasting processing resources. I'm experimenting on a way to isolate regions that have near constant motion and tune them out. I did not find a way to do this with generic background subtraction since it considers objects that move to be foreground. My efforts to tune the detection to include this kind of motion as background made foreground detection worse (for obvious reasons).
I'm now working on an approach that will use the current background/foreground detection as is but implement another layer that will create a **dynamic mask** for regions with near-constant motion to then ignore. When the motion stops the area is no longer masked. My initial tests are promising but it still needs some work. My hope is that this will remove the need to manually mask areas of the frame on a permanent basis. It should also help with situations like spider webs and wind storms that pop up and trigger a lot of false alerts.

The other issue I'm having is where a moving object overlaps a stationary object and causing the stationary object to act as new/moving even though it's not. This dynamic mask will not solve that problem. Instead I'm thinking adding different feature would be better, something like periodic object checks of the whole frame to identify objects that exist while no motion is detected. When a "new" object is detected via motion then we can reference the stationary object list for things that should be ignored. I'm not sure if that is a good solution or not, but I'll explore it later.

**Comment by tve** [comment]
As far as I can tell, the motion detection in Frigate is a slight variation on what's described in https://www.pyimagesearch.com/2015/05/25/basic-motion-detection-and-tracking-with-python-and-opencv/
It keeps an average of the last 30 frames as background.

I'm also having quite some issues with wind motion. I don't really like the idea of dynamic masks though because in my case the real object motion can easily happen in front of that wind motion. One thing I would love is for those motion bounding boxes to show some of the parameter values so I could perhaps begin by simply tuning the existing motion detector...

**Reply by gpete** [reply]
Hey, thanks for the input! I'm glad to have someone to bounce ideas off of. I have a test script I've been using that implements the meat of the Frigate motion detection while allowing me to test out ideas like a dynamic mask. If you're interested I could clean it up a bit for sharing and you could try out some of your ideas with your own cameras and test clips.

I love the power of the dynamic mask but you are right about foreground objects getting detected too. A common example in my setup is someone sitting on the porch or patio reading or a visitor standing and having a long conversation outside in front of the camera. The proof-of-concept I've been toying with has parameters that allow you to specify how quickly pixels are added to the mask and the rate of decay where it is removed from the mask. Added too slowly and it's not useful enough, too quick and it picks up foreground objects. The decay also has to be tuned as too slow of a decay masks erroneous regions too long, too quickly and the mask will drop the useful regions.

Another thought I had could utilize a more persistent and long-term approach. The goal would be having a mask that adapts to plant growth and seasonal change in a time frame that makes sense to them: days and weeks instead of seconds, minutes, or hours. The mask could be stored off hourly and reloaded on restarts. That also makes it possible to view it and manually manipulate it for testing. Definitely easier said than done but I've got to start with a good idea first.

**Reply by tve** [reply]
I think it's useful to take a step back here and ask what we're trying to accomplish... I hadn't seen `/api/stats` but now that I look at it, I'm close to being detection bound 'cause I'm waiting on an EdgeTPU to be delivered, but once I have that I should have ~3x headroom in terms of detections.
I see the motion detection serving two purposes: not having to send each frame to the object detector, and creating a bounding box so what is sent to the detector is well framed.
When we're talking about "improve BG subtraction" I believe the purpose would be to reduce the number of detections run? If so, what other options are there and how important is it really?
One idea that comes to mind is that the wind motion bounding boxes seem to be quite consistent, maybe we could prioritize detections to run on novel bounding boxes? 

BTW, your test script sounds very interesting! I was pondering a notion of "replay" where one could replay a saved recording snippet in frigate and see how it fares with a different config. But I suspect a script is more flexible in the long run.

**Reply by tve** [reply]
WRT script... Have you seen https://github.com/blakeblackshear/frigate/blob/master/frigate/process_clip.py#L150-L155 ?
Dunno what it exactly does nor how to use it, but definitely looks like a way to process a video clip for test/debug purposes...
